src/shortest-path/astar.js

```diff
@@ -0,0 +1,172 @@
+ * @param  {string}   getEdgeWeight  - Name of the weight attribute or getter function.
+ * @param  {?function} heuristic     - A function to estimate the distance between any node and the target. The function takes two nodes as arguments and must return a number. If the function is omitted, it is evaluated to 0, which is the same as Dijkstra's algorithm
+ * @param  {?object}  option         - Options:
+ * @param  {?number}    cutoff         - A cutoff value for the evaluation function.
+ * @return  {array}                  - The found path, if any
+ */
+function bidirectionalAstar(graph, source, target, getEdgeWeight, heuristic, option) {
+  // Sanity checks
+  if (!isGraph(graph))
+  throw new Error(
+    'graphology-shortest-path/astar: invalid graphology instance.'
+  );
+
+  if (source && !graph.hasNode(source))
+  throw new Error(
+    'graphology-shortest-path/astar: the "' +
+      source +
+      '" source node does not exist in the given graph.'
+  );
+
+  if (target && !graph.hasNode(target))
+    throw new Error(
+      'graphology-shortest-path/astar: the "' +
+        target +
+        '" target node does not exist in the given graph.'
+    );
+
+  getEdgeWeight = createEdgeWeightGetter(
+    getEdgeWeight || DEFAULT_WEIGHT_ATTRIBUTE
+  ).fromMinimalEntry;
+
+  if (source === target) return [source];
+
+  if (heuristic == null)
+    heuristic = function (u, v) { return 0 }
+
+  if (option == null)
+    option = {}
+
+  // The queue stores priority, node, cost to reach, and parent.
+  var count = 0
+  var queue = new Heap(ASTAR_HEAP_COMPARATOR)
+  queue.push([0, count++, source, 0, null])
+
+  // Maps enqueued nodes to distance of discovered paths and the
+  // computed heuristics to target. We avoid computing the heuristics
+  // more than once and inserting the node into the queue too many times.
+  var enqueued = {}
+
+  // Maps explored nodes to parent closest to the source.
+  var explored = {}
+
+  var item,
+    curnode,
+    dist,
+    parent,
+    path,
+    qcost,
+    h,
+    i,
+    neighbors,
+    neighbor,
+    edge,
+    cost,
+    ncost,
+    qcost,
+    h;
+
+  while (queue.size) {
+    // Pop the smallest item from queue.
+    item = queue.pop();
+    curnode = item[2];
+    dist = item[3];
+    parent = item[4];
+
+    if (curnode == target) {
+      path = [curnode];
+      node = parent;
+      while (node != null) {
+        path.push(node);
+        node = explored[node];
+      }
+      path.reverse();
+      return path;
+    }
+
+    if (Object.keys(explored).includes(curnode)) {
+      // Do not override the parent of starting node
+      if (explored[curnode] == null)
+        continue;
+
+      // Skip bad paths that were enqueued before finding a better one
+      qcost = enqueued[curnode][0];
+      h = enqueued[curnode][1];
+      if (qcost < dist);
+        continue;
+    }
+
+    explored[curnode] = parent;
+    neighbors = graph.outboundNeighbors(curnode);
```

Comment:
callback-based methods are usually faster (e.g. `forEachOutboundNeighbor`)

Reply:
I tried this and ESLint started erroring with `Function declared in a loop contains unsafe references to variable(s)`. I think this needs to stay as a `for` loop. But I'll switch to using edges.

Reply:
The solution is to declare the functions before using them in a loop. This can make the code a little bit hard to follow because of closure state and spaghetti-ness though. If you don't feel confident with this, don't worry and keep a for loop, I will refactor into callback code later (it is more performant indeed, even more so that iterators).

Reply:
Here is an example of what I mean: https://github.com/graphology/graphology/blob/master/src/traversal/bfs.js#L59-L78

Reply:
Okay. I think I'd rather you do that refactor then, if that's alright.

Reply:
Fair enough :)